File: src/lib/tool-manifest/tools/list-layers.ts
import { z } from 'zod';
import { useEditorStore } from '@/store';
import type { ToolManifest } from '../types';

const input = z.object({}).describe('No input.');

const output = z.object({
  layers: z.array(
    z.object({
      id: z.string(),
      type: z.string(),
      name: z.string(),
      isActive: z.boolean(),
      adjustmentCount: z.number(),
    }),
  ),
});

export const listLayersTool: ToolManifest<typeof input, typeof output> = {
  name: 'list_layers',
  kind: 'query',
  description:
    'List the layers in the current document. Most documents have a single image layer; this matters when the LLM needs to apply different adjustments to different layers.',
  inputSchema: input,
  outputSchema: output,
  handler: () => {
    const s = useEditorStore.getState();
    return {
      layers: s.layers.map((l) => ({
        id: l.id,
        type: l.type,
        name: l.name ?? l.type,
        isActive: l.id === s.activeLayerId,
        adjustmentCount: l.adjustmentStack?.adjustments.length ?? 0,
      })),
    };
  },
};

File: backend/app/tools/atomic/list_layers.py
from __future__ import annotations

from pydantic import BaseModel, Field

from app.state.document import SessionDocument
from app.tools.base import BackendTool, ToolPermissions


class _Input(BaseModel):
    pass


class _LayerSummary(BaseModel):
    id: str
    type: str
    name: str
    is_active: bool
    adjustment_count: int


class _Output(BaseModel):
    layers: list[_LayerSummary] = Field(default_factory=list)


class ListLayersTool(BackendTool[_Input, _Output]):
    name = "list_layers"
    kind = "query"
    description = (
        "List the layers in the current document. Most documents have a single "
        "image layer."
    )
    input_schema = _Input
    output_schema = _Output
    permissions = ToolPermissions(requires_image=False)

    async def handler(self, doc: SessionDocument, input: _Input) -> _Output:  # noqa: A002
        return _Output(layers=[
            _LayerSummary(
                id="l_image", type="image", name="Background",
                is_active=True,
                adjustment_count=sum(1 for w in doc.widgets.values() if w.status == "active"),
            )
        ])
